session/kvdb_store.go

```diff
@@ -444,6 +421,99 @@ func (db *BoltStore) listSessions(filterFn func(s *Session) bool) ([]*Session,
 	return sessions, nil
 }
 
+// DeleteReservedSessions deletes all sessions that are in the StateReserved
+// state.
+//
+// NOTE: this is part of the Store interface.
+func (db *BoltStore) DeleteReservedSessions() error {
+	return db.Update(func(tx *bbolt.Tx) error {
+		sessionBucket, err := getBucket(tx, sessionBucketKey)
+		if err != nil {
+			return err
+		}
+
+		return sessionBucket.ForEach(func(k, v []byte) error {
+			// We'll also get buckets here, skip those (identified
+			// by nil value).
+			if v == nil {
+				return nil
+			}
+
+			session, err := DeserializeSession(bytes.NewReader(v))
+			if err != nil {
+				return err
+			}
+
+			if session.State != StateReserved {
+				return nil
+			}
+
+			err = sessionBucket.Delete(k)
+			if err != nil {
+				return err
+			}
+
+			idIndexBkt := sessionBucket.Bucket(idIndexKey)
+			if idIndexBkt == nil {
+				return ErrDBInitErr
+			}
+
+			// Delete the entire session ID bucket.
+			err = idIndexBkt.DeleteBucket(session.ID[:])
+			if err != nil {
+				return err
+			}
+
+			groupIdIndexBkt := sessionBucket.Bucket(groupIDIndexKey)
+			if groupIdIndexBkt == nil {
+				return ErrDBInitErr
+			}
+
+			groupBkt := groupIdIndexBkt.Bucket(session.GroupID[:])
+			if groupBkt == nil {
+				return ErrDBInitErr
+			}
+
+			sessionIDsBkt := groupBkt.Bucket(sessionIDKey)
+			if sessionIDsBkt == nil {
+				return ErrDBInitErr
+			}
+
+			var (
+				seqKey      []byte
+				numSessions int
+			)
+			err = sessionIDsBkt.ForEach(func(k, v []byte) error {
+				numSessions++
+
+				if !bytes.Equal(v, session.ID[:]) {
+					return nil
```

Comment:
Thought this might be worth pointing out for @jamaljsr, so just leaving a comment here, as this always confuses me with `kvdb` buckets `ForEach` :). I.e. there's nothing here to address, this looks good but is a bit confusing :rocket:!

The `return nil` doesn't break the `ForEach` loop here, but rather continues the iteration with the next element of the bucket. Only non-nil errors actually breaks the loop.